### ISSUE TITLE
stencil-ds-vue-template

### ISSUE THREAD

**Issue by elwinvaneede** [ionic-team/stencil-ds-output-targets#77]
Thanks for the new Vue Output Target! I look forward in using it.

But I haven't been able to get it working properly with my Stencil component library testing repo: https://github.com/elwinvaneede/elwins-test-web-components. Unlike the React & Angular output targets.

The docs say:

> There is an example component library package available on Github so that you can get started.`

But I can't find the `stencil-ds-vue-template` repo (like https://github.com/ionic-team/stencil-ds-react-template & https://github.com/ionic-team/stencil-ds-angular-template), so I've had to guess the correct `package.json` setup. 

I've also found the [example-project](https://github.com/ionic-team/stencil-ds-output-targets/tree/master/packages/example-project) in this repo, which seem like new example templates, but those `package.json` files seem to have unnecessary devDependencies like Rollup, and too many regular dependencies?

This is the one I currently use for my Vue output target: https://github.com/elwinvaneede/elwins-test-web-components-vue/blob/master/package.json.

But when using it in a Vue project created with the `vue-cli`, my components don't properly render. Any styling and functionality is missing. Only the text I add in a slot is visible in my browser. I can inspect the component in Chrome (where any props I've defined are not rendered). I don't get any console errors, and the Vue DevTools also show the components (with the props I've defined).

Do you have ideas how to fix this, and are there any plans on adding a `stencil-ds-vue-template` repo?

Thanks in advance.

**Comment by dtieber** [comment]
Same issue here. Any news on this?

**Comment by elwinvaneede** [comment]
@jthoms1 the current Vue output-target seems to be based on Vue 2, but my understanding is that Ionic Vue 1.0 will be based on Vue 3. Will this output-target be eventually based on Vue 3? Or will it support both versions?

**Comment by robaxelsen** [comment]
@jthoms1 do you have any input on this?

**Comment by elwinvaneede** [comment]
With the new `0.0.3` release of `@stencil/vue-output-target`, I'm still having the same issue. And it seems I'm not alone.
I read in the recent contributors chat notes that Josh is currently in another team (https://github.com/ionic-team/stencil/blob/master/notes/2020-08-12.md). 

@adamdbradley could you shed some light on this? I would love to start using Stencil for a Design System within my company, but it's fairly critical that the components offer the same DX as regular Vue components.

Thanks!

**Comment by robaxelsen** [comment]
I hear you're working in Ionic Vue (3) atm, @liamdebeasi. As @jthoms1 does not seem to work on this any longer, and hasn't responded, is this something you can look into?

**Comment by liamdebeasi** [comment]
The output targets are currently being updated to support Vue 3/fix bugs. Still in progress, so hang tight - but we do plan to publish a stable version soon. Thanks!

**Comment by elwinvaneede** [comment]
Great, I'm looking forward to it, thanks for the update @liamdebeasi

**Comment by robaxelsen** [comment]
> The output targets are currently being updated to support Vue 3/fix bugs. Still in progress, so hang tight - but we do plan to publish a stable version soon. Thanks!

Great, thanks @liamdebeasi! Appreciate the news and effort a lot :heart:

Is this on any roadmap, or has a specific ETA (we have projects depending on this, so just want to understand if we're talking 3 weeks or 3+ months, hehe).

**Comment by liamdebeasi** [comment]
Soon 👀 . We plan on having a build of Ionic Vue late summer/early fall, so the Stencil Vue Output Targets should be published around then as well.

**Comment by bjankord** [comment]
Just wanted to say I'm seeing the same issue as @elwinvaneede 

> when using it in a Vue project created with the vue-cli, my components don't properly render. Any styling and functionality is missing. Only the text I add in a slot is visible in my browser.

I'm using "@stencil/vue-output-target": "0.0.3" currently to generate the vue components. 

My current plan is to just use the web components directly in Vue. Theres not too much difference since Vue has such great support for web components, however it would be nice to have them converted to Vue components automagically via a stencil output target.

I saw in there is a vue3 branch in development: https://github.com/ionic-team/stencil-ds-output-targets/tree/vue3

I'm curious what the plans are for the Vue output target. Will we be seeing a Vue 2 and Vue 3 output target for stencil?

**Comment by robaxelsen** [comment]
> Just wanted to say I'm seeing the same issue as @elwinvaneede
> 
> > when using it in a Vue project created with the vue-cli, my components don't properly render. Any styling and functionality is missing. Only the text I add in a slot is visible in my browser.
> 
> I'm using "@stencil/vue-output-target": "0.0.3" currently to generate the vue components.
> 
> My current plan is to just use the web components directly in Vue. Theres not too much difference since Vue has such great support for web components, however it would be nice to have them converted to Vue components automagically via a stencil output target.
> 
> I saw in there is a vue3 branch in development: https://github.com/ionic-team/stencil-ds-output-targets/tree/vue3
> 
> I'm curious what the plans are for the Vue output target. Will we be seeing a Vue 2 and Vue 3 output target for stencil?

@bjankord I believe your question was already answered by @liamdebeasi above:

> The output targets are currently being updated to support Vue 3/fix bugs. Still in progress, so hang tight - but we do plan to publish a stable version soon. Thanks!

**Comment by bjankord** [comment]
One thing I'm curious about is if Vue 2 will be supported whenever the stencil-ds-output-target for Vue becomes available/stable.

**Comment by liamdebeasi** [comment]
Moving forward, `@stencil/vue-output-target` will support Vue 3+ only. Vue 2 will not be supported.

**Comment by spaceninja** [comment]
We were able to work around this bug using v0.0.3 of the Vue output target by adding the following code to `main.js` in our Vue 2 app:

```js
import { defineCustomElements, applyPolyfills } from 'vue-component-lib';
applyPolyfills().then(() => defineCustomElements());
```

Where `vue-component-lib` is the package you make from the generated files from your Vue output target.

**Comment by elwinvaneede** [comment]
Thanks @spaceninja.

I've also found out that you can add the following options to the vueOutputTarget config in the `stencil.config.ts` (which I also needed to do for the React Output Target):

```
  includeDefineCustomElements: true,
  includePolyfills: true,
```
(see https://github.com/elwinvaneede/elwins-test-web-components/blob/master/stencil.config.ts#L21)

With this, you don't have to add the code above from @spaceninja to the `main.js` of a Vue 2 app (because this code is now added to the Vue Framework Binding) and also fixes it.

**Comment by robaxelsen** [comment]
@liamdebeasi any update on this?

Also, I'd be curious to know how you've dealt with the [seemingly broken deprecation of .prop on attributes](https://github.com/vuejs/vue-next/blob/d4527230e40c4728e5becdd35c3e039f0992ae4c/packages/compiler-core/src/transforms/vBind.ts#L13) introduced in Vue 3. We are not able to pass attribute as object anymore for Stencil.js web component with Vue 3, but works when we try Vue 3 with "hand rolled" web component. Any insight?

**Comment by robaxelsen** [comment]
> @liamdebeasi any update on this?
> 
> Also, I'd be curious to know how you've dealt with the [seemingly broken deprecation of .prop on attributes](https://github.com/vuejs/vue-next/blob/d4527230e40c4728e5becdd35c3e039f0992ae4c/packages/compiler-core/src/transforms/vBind.ts#L13) introduced in Vue 3. We are not able to pass attribute as object anymore for Stencil.js web component with Vue 3, but works when we try Vue 3 with "hand rolled" web component. Any insight?

Made a reproducer: https://github.com/ionic-team/stencil/issues/2701

**Comment by elwinvaneede** [comment]
By looking at the example-project in the [vue3 branch](https://github.com/ionic-team/stencil-ds-output-targets/tree/vue3), and combining it with the [source code of the Ionic Vue package](https://github.com/ionic-team/ionic-framework/tree/master/packages/vue) I got it working for [my test web component library](https://github.com/elwinvaneede/elwins-test-web-components).

I've made a PR for it: https://github.com/ionic-team/stencil-ds-output-targets/pull/127

**Comment by robaxelsen** [comment]
> By looking at the example-project in the [vue3 branch](https://github.com/ionic-team/stencil-ds-output-targets/tree/vue3), and combining it with the [source code of the Ionic Vue package](https://github.com/ionic-team/ionic-framework/tree/master/packages/vue) I got it working for [my test web component library](https://github.com/elwinvaneede/elwins-test-web-components) with Vue 3.
> 
> I've made a PR for it: #127

Nice @elwinvaneede! Love the work you've put into this. Does this also solve passing of non-primitive props (.prop was deprecated in Vue 3) and the Vue 3 events changes?

**Comment by elwinvaneede** [comment]
> > By looking at the example-project in the [vue3 branch](https://github.com/ionic-team/stencil-ds-output-targets/tree/vue3), and combining it with the [source code of the Ionic Vue package](https://github.com/ionic-team/ionic-framework/tree/master/packages/vue) I got it working for [my test web component library](https://github.com/elwinvaneede/elwins-test-web-components) with Vue 3.
> > I've made a PR for it: #127
> 
> Nice @elwinvaneede! Love the work you've put into this. Does this also solve passing of non-primitive props (.prop was deprecated in Vue 3) and the Vue 3 events changes?

Thanks, I think so, because Ionic Vue uses the same output target, but I haven't tried it out yet. Will let you know if it works.

**Comment by robaxelsen** [comment]
> Thanks, I think so, because Ionic Vue uses the same output target, but I haven't tried it out yet. Will let you know if it works.

@elwinvaneede did you have any luck? This might be of interest:
https://github.com/vuejs/vue-next/issues/2343
https://github.com/ionic-team/stencil/issues/2701

**Comment by elwinvaneede** [comment]
@robaxelsen I just created a Stringify test component (slightly based on your reproducer), and I've got the same problem unfortunately. If you go to https://codesandbox.io/s/stencil-vue3-output-target-prop-test-5epoy it doesn't display a stringified array (btw with version 0.3.3 you get the "[object Object],[object Object]" output you used in your reproducer). But if you then change the value of the data prop to something like `test1` and then back to `test` it does work.

My guess it has to do with the fact that `defineCustomElements()` is async, and Vue doesn't wait on it finishing. And since Vue 3 is now first trying to bind to a property and then falls back to an attribute (instead of explicitly having to use `.prop` with Vue 2), results in this problem.

I think Ionic Vue solves this with a [IonicPlugin](https://github.com/ionic-team/ionic-framework/blob/49da5f85e3be3dc31777939f41b4d16dca0b926b/packages/vue/src/ionic-vue.ts), that you need to use like `createApp(App).use(IonicVue)`.

**Comment by elwinvaneede** [comment]
Didn't help either: https://codesandbox.io/s/stencil-vue-output-target-prop-test-with-plugin-j7jj4

Maybe Ionic Vue doesn't suffer from this because of this in the main.ts:

```
router.isReady().then(() => {
  app.mount('#app');
});
```

Hopefully someone from the Ionic Team responds.

**Comment by elwinvaneede** [comment]
@robaxelsen I got it working with this code in the main.js:

```js
import { createApp } from "vue";
import App from "./App.vue";
import { defineCustomElements } from "elwins-test-web-components/loader";

const app = createApp(App);

defineCustomElements(window).then(() => {
  app.mount("#app");
});
```
See https://codesandbox.io/s/stencil-vue3-output-target-wait-for-define-custom-elements-ycfth

**Comment by robaxelsen** [comment]
> @robaxelsen I got it working with this code in the main.js (remove the applyPolyfills if not needed):
> 
> ```js
> import { createApp } from "vue";
> import App from "./App.vue";
> import { applyPolyfills, defineCustomElements } from "elwins-test-web-components/loader";
> 
> const app = createApp(App);
> 
> applyPolyfills().then(() => {
>   defineCustomElements().then(() => {
>     app.mount("#app");
>   });
> });
> ```
> 
> See https://codesandbox.io/s/stencil-vue3-output-target-wait-for-define-custom-elements-ycfth

@elwinvaneede Awesome! Well done.

@liamdebeasi Is this something that could be used to finish the official vue 3 output target? Or is there already something in the works that we can try out soon?

**Comment by liamdebeasi** [comment]
There isn't any good solution right now because `defineCustomElements` is async as noted in https://github.com/ionic-team/stencil-ds-output-targets/issues/77#issuecomment-723467192. Right now Ionic Vue waits for the Vue Router to be ready (`await router.isReady()`) before mounting.

**Comment by elwinvaneede** [comment]
I've incorporated it in my PR to improve the docs.

**Comment by andrew9994** [comment]
Thanks @elwinvaneede for your code, was really helpful for our Vue3 output target library! 👍

**Comment by andrew9994** [comment]
@elwinvaneede , i just have a question regarding the code: is it possible to listen on events without defining them `componentOptions` ? Because currently they get mixed up with the props in the `componentProps`
